### PR TITLE
[TRAFODION-2891] fix the bufoverrun Critical error checked by TScanCode

### PR DESCRIPTION
**Pull request by xiaozhongwang**
fix the bufoverrun Critical error checked by TScanCode

### CODE REVIEW

File: core/conn/odb/src/odb.c [hunk 1]
```diff
@@ -5313,7 +5313,7 @@ static void etabadd(char type, char *run, int id)
                     }
                 }
                 if ( etab[no].type == 'e' ) { /* name & create output file */
-                    for ( i = j = 0; etab[no].tgt[i] && i < sizeof(buff); i++ ) {
+                    for ( i = j = 0; i < sizeof(buff) && etab[no].tgt[i]; i++ ) {
```
Comment (traflm) [inline core/conn/odb/src/odb.c]: I don't really get this fully understand.
If i is out of bound for arry tgt, why this change will prevent the tag[i] to be executed?

File: core/conn/odbc/src/odbc/Common/linux/sqmem.cpp [hunk 2]
```diff
@@ -400,7 +400,7 @@ void TestPool(void * membase, int length)
    long j, index;
    short error;
    long pass;
-   void * pool_ptrs[256];
+   void * pool_ptrs[256 + 1];
```
Comment (traflm) [inline core/conn/odbc/src/odbc/Common/linux/sqmem.cpp]: very good catch!

File: core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp [hunk 3]
```diff
@@ -5153,7 +5153,7 @@ odbc_SQLSvc_GetSQLCatalogs_sme_(
                  {
                     ERROR_DESC_def *p_buffer = QryCatalogSrvrStmt->sqlError.errorList._buffer;
                     strncpy(RequestError, p_buffer->errorText,sizeof(RequestError) -1);
-                    RequestError[sizeof(RequestError)] = '\0';
+                    RequestError[sizeof(RequestError) - 1] = '\0';
```
Comment (traflm) [inline core/conn/odbc/src/odbc/nsksrvrcore/srvrothers.cpp]: very good catch!

File: core/sqf/src/seatrans/tm/hbasetmlib2/hbasetm.h [hunk 4]
```diff
@@ -139,9 +139,9 @@ class CHbaseTM : public JavaObjectInterfaceTM
       JM_REGTRUNCABORT,
       JM_DROPTABLE,
       JM_RQREGINFO,
-      JM_LAST
+      JM_TMLAST
    };
-   JavaMethodInit JavaMethods_[JM_LAST];
+   JavaMethodInit JavaMethods_[JM_TMLAST];
```
Comment (traflm) [inline core/sqf/src/seatrans/tm/hbasetmlib2/hbasetm.h]: This is critical!

File: core/conn/odb/src/odb.c (hunk 1 above)
Comment (xiaozhongwang) [reply core/conn/odb/src/odb.c]: Most of time, they are equivalent.  There are exception in a special case 
Most of compiler judge the condition from left to right, so if the tgt is at the end of memory, and the offset i is equal to sizeof(buff), the access to memory goes beyond memory.
This will make a coredump, and cann't be repeatable.

File: core/sqf/src/seatrans/tm/hbasetmlib2/hbasetm.h (hunk 4 above)
Comment (xiaozhongwang) [reply core/sqf/src/seatrans/tm/hbasetmlib2/hbasetm.h]: There is  another define of JM_LAST in this file, and their value is different.

File: core/conn/odb/src/odb.c (hunk 1 above)
Comment (DaveBirdsall) [reply core/conn/odb/src/odb.c]: I tried to make sense of this too. It looks like etab[no].tgt is populated in function parseopt from some string buffer; I can't tell if it is buff (which is a static variable declared as a char[256]) or something else. My suspicion is that the test "i < sizeof(buff)" has nothing to do with etab[no].tgt but rather some other buffer being copied to. In any case, reversing the order of these tests seems harmless.
